File: study/2022/22.09.20/HHS/파괴되지않은건물.java
import java.util.*;

class Solution {
    
    static int answer;
    static int R, C;
    static int [][] cal;
    
    public int solution(int[][] board, int[][] skill) {
        R = board.length;
        C = board[0].length;
        cal = new int[R+1][C+1];
        
        for(int t = 0; t < skill.length; t++){
            int r1 = skill[t][1];
            int c1 = skill[t][2];
            int r2 = skill[t][3];
            int c2 = skill[t][4];
            int d = skill[t][5];
            if(skill[t][0] == 1) //적군
                d = -d;
            
            cal[r1][c1] += d;
            cal[r2+1][c2+1] += d;
            cal[r2+1][c1] -= d;
            cal[r1][c2+1] -= d;
        }
        
        for(int i = 1; i < R; i++){
            for(int j = 0; j < C; j++){
                cal[i][j] += cal[i-1][j];
            }
        }
        for(int j = 1; j < C; j++){
            for(int i = 0; i < R; i++){
                cal[i][j] += cal[i][j-1];
            }
        }
        
        for(int i = 0; i < R; i++){
            for(int j = 0; j < C; j++){
                if(board[i][j] + cal[i][j] > 0)
                    answer++;
            }
        }
        
        return answer;
    }
}

File: study/2022/22.09.20/EJH/파괴되지 않은 건물.py
def solution(board, skill):
    answer = 0
    N = len(board)
    M = len(board[0])
    imos = [[0 for _ in range(M + 1)] for _ in range(N + 1)]
    for s in skill:
        t, r1, c1, r2, c2, degree = s
        if t == 1:
            degree *= -1

        imos[r1][c1] += degree
        imos[r1][c2 + 1] -= degree
        imos[r2 + 1][c1] -= degree
        imos[r2 + 1][c2 + 1] += degree

    for r in range(N + 1):
        for c in range(1, M + 1):
            imos[r][c] += imos[r][c - 1]

    for c in range(M + 1):
        for r in range(1, N + 1):
            imos[r][c] += imos[r - 1][c]

    for r in range(N):
        for c in range(M):
            if board[r][c] + imos[r][c] > 0:
                answer += 1

    return answer


print(solution([[5, 5, 5, 5, 5], [5, 5, 5, 5, 5], [5, 5, 5, 5, 5], [5, 5, 5, 5, 5]],
               [[1, 0, 0, 3, 4, 4], [1, 2, 0, 2, 3, 2], [2, 1, 0, 3, 1, 2], [1, 0, 1, 3, 3, 1]]))
print(solution([[1, 2, 3], [4, 5, 6], [7, 8, 9]], [[1, 1, 1, 2, 2, 4], [1, 0, 0, 1, 1, 2], [2, 2, 0, 2, 0, 100]]))
